### PR TITLE
Fixing constants in GROUP BY keys with analyzer (again)

### LINKED ISSUE #65468: Cannot find column _CAST with allow_experimental_analyzer
24.6.1

```sql
CREATE TABLE t (hr DateTime, ts DateTime) ENGINE=Memory
as select '2000-01-01' d, d;
```

Q1) Not distributed table + analyzer https://fiddle.clickhouse.com/7e58d62f-4442-4a1a-ae49-20a38a0c649b
expected result, all OK:

```sql
SELECT
  count(),
  tuple(nullIf(toDateTime(formatDateTime(hr, '%F %T', 'America/Los_Angeles'), 'America/Los_Angeles'), toDateTime(0)))  as c1,
  defaultValueOfArgumentType(toTimeZone(ts, 'America/Los_Angeles'))  as c2,
  formatDateTime(hr, '%F %T', 'America/Los_Angeles')  as c3
FROM   t
GROUP BY c1, c2, c3
settings allow_experimental_analyzer=1 
format Pretty;

   +---------+-------------------------+---------------------+---------------------+
   | count() | c1                      |                  c2 | c3                  |
   +---------+-------------------------+---------------------+---------------------+
1. |       1 | ('1999-12-31 16:00:00') | 1969-12-31 16:00:00 | 1999-12-31 16:00:00 |
   +---------+-------------------------+---------------------+---------------------+
```

Q2) distributed table + analyzer  https://fiddle.clickhouse.com/e8a9dfea-1ff5-4614-a4ef-8449cfda66f3

```sql
SELECT
  count(),
  tuple(nullIf(toDateTime(formatDateTime(hr, '%F %T', 'America/Los_Angeles'), 'America/Los_Angeles'), toDateTime(0)))  as c1,
  defaultValueOfArgumentType(toTimeZone(ts, 'America/Los_Angeles'))  as c2,
  formatDateTime(hr, '%F %T', 'America/Los_Angeles')  as c3
FROM   remote('127.0.0.{1,2}', default,  t)
GROUP BY c1, c2, c3
settings allow_experimental_analyzer=1;

DB::Exception: THERE_IS_NO_COLUMN
```

Q3)  Not distributed table + No analyzer https://fiddle.clickhouse.com/ca357000-b2c9-4897-8468-4d32bcad2062

```sql
SELECT
   count(),
   tuple(nullIf(toDateTime(formatDateTime(hr, '%F %T', 'America/Los_Angeles'), 'America/Los_Angeles'), toDateTime(0)))  as c1,
   defaultValueOfArgumentType(toTimeZone(ts, 'America/Los_Angeles'))  as c2,
   formatDateTime(hr, '%F %T', 'America/Los_Angeles')  as c3
FROM t
GROUP BY c1, c2, c3
settings allow_experimental_analyzer=0;

DB::Exception: Column `ts` is not under aggregate function and not in GROUP BY
```

----

Probably it's another/known issue https://fiddle.clickhouse.com/fa39a068-cfc1-4eab-8d7f-c848c5ba9092

```sql
CREATE TABLE t (hr DateTime, ts DateTime) ENGINE=Memory
as select '2000-01-01' d, d;

SELECT
     count(),
     now() as c1
FROM remote('127.0.0.{1,2}', default,  t)
GROUP BY c1
settings allow_experimental_analyzer=1;

DB::Exception: Cannot find column `_CAST(1718827653_DateTime
```

### PR DESCRIPTION
**Pull request by KochetovNicolai**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix `Cannot find column` error for queries with constant expression in `GROUP BY` key and new analyzer enabled.

Fixes #65468 (which was broken after #64519)

### CODE REVIEW

File: src/Analyzer/Resolve/QueryAnalyzer.cpp [hunk 1]
```diff
@@ -3410,14 +3410,14 @@ ProjectionNames QueryAnalyzer::resolveFunction(QueryTreeNodePtr & node, Identifi
...
-            function_name == "shardCount" || function_name == "hostName" || function_name == "tcpPort";
+        // bool disable_constant_folding = function_name == "__getScalar" || function_name == "shardNum" ||
+        //     function_name == "shardCount" || function_name == "hostName" || function_name == "tcpPort";
 
         /** If function is suitable for constant folding try to convert it to constant.
           * Example: SELECT plus(1, 1);
           * Result: SELECT 2;
           */
-        if (function_base->isSuitableForConstantFolding() && !disable_constant_folding)
+        if (function_base->isSuitableForConstantFolding()) // && !disable_constant_folding)
```
Comment (KochetovNicolai) [inline src/Analyzer/Resolve/QueryAnalyzer.cpp:3420]: Currently, I want to check what tests it breaks.

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit 8c2e4900e301b7b54c8c266f2c2c8ef4a4ef5e2c with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/66433/8c2e4900e301b7b54c8c266f2c2c8ef4a4ef5e2c/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Style check</td><td>Runs a set of checks to keep the code style clean. If some of tests failed, see the related log from the report</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/66433/8c2e4900e301b7b54c8c266f2c2c8ef4a4ef5e2c/style_check.html">❌ failure</a></td></tr>
<tbody>
</table>

File: src/Planner/PlannerExpressionAnalysis.cpp [hunk 2]
```diff
@@ -85,14 +85,14 @@ bool canRemoveConstantFromGroupByKey(const ConstantNode & root)
         else if (function_node)
         {
             /// Do not allow removing constants like `hostName()`
-            if (!function_node->getFunctionOrThrow()->isDeterministic())
+            if (function_node->getFunctionOrThrow()->isServerConstant())
```
Comment (novikd) [inline src/Planner/PlannerExpressionAnalysis.cpp:88]: Will it work for `random` functions?

Comment (KochetovNicolai) [reply src/Planner/PlannerExpressionAnalysis.cpp:88]: Random does not return constant result, so no constant folding.
For `randConst` we constant fold and substitute result to constant and remove it from GROUP BY key.
```
SELECT
    count(),
    randConstant() AS c1
FROM remote('127.0.0.{3,2}', system, one)
GROUP BY c1
SETTINGS allow_experimental_analyzer = 1

Query id: 15fef62f-a79a-4088-a3b2-44bf5259b2b4

   ┌─count()─┬────────c1─┐
1. │       2 │ 563172653 │
   └─────────┴───────────┘

```

The issue is only for special functions that are not constant on the initiator (so they are not substituted to constants) but constant on the replica and removed from the GROUP BY key.

**Pull request merged**